File: exam/src/services/validator/ProposalValidator.ts
import { DocumentValidator } from "./DocumentValidator";
import { Document } from "../../models/Document";

export class ProposalValidator extends DocumentValidator {
    private static readonly MAX_SIZE_MB = 5;
    private static readonly ALLOWED_EXTENSIONS = [".pdf", ".docx"];
    private static readonly REQUIRED_METADATA = ["proposalDate", "client"];

    protected validateFileSize(document: Document): void {
        if (document.getFileSize() > this.MAX_SIZE_MB * 1024 * 1024) {
            throw new Error(`Validation Error: Proposal file size exceeds ${this.MAX_SIZE_MB}MB.`);
        }
    }

    protected validateFileExtension(document: Document): void {
        if (!this.ALLOWED_EXTENSIONS.includes(document.getExtension())) {
            throw new Error(`Validation Error: Invalid extension. Allowed: ${this.ALLOWED_EXTENSIONS.join(", ")}`);
        }
    }

    protected validateMetadata(document: Document): void {
        for (const meta of this.REQUIRED_METADATA) {
            if (!document.getMetadata() || !document.getMetadata()[meta]) {
                throw new Error(`Validation Error: Missing required metadata field '${meta}'.`);
            }
        }
    }
}

File: exam/src/validator/documentValidators/ProposalValidator.java
package validator.documentValidators;

import document.Document;
import validator.DocumentValidator;

public class ProposalValidator extends DocumentValidator {

    public ProposalValidator() {
        super();
    }
    
    @Override
    protected boolean validateSize(Document document) {
        return document.getSize() <= 5;
    }

    @Override
    protected boolean validateExtension(Document document) {
        if (isDocx(document)) return true;
        if (isPdf(document)) return true;
        return false;
    }
    
    private boolean isDocx(Document document) {
        return document.getExtension().equals("docx");
    }
    
    private boolean isPdf(Document document) {
        return document.getExtension().equals("pdf");
    }

    @Override
    protected boolean validateMetadata(Document document) {
        if (!metadataContainsProposalDate(document)) return false;
        if (!metadataContainsClient(document)) return false;
        return true;
    }

    private boolean metadataContainsProposalDate(Document document) {
        return document.getMetadata().containsKey("proposalDate");
    }

    private boolean metadataContainsClient(Document document) {
        return document.getMetadata().containsKey("client");
    }
}
